sklearnex/neighbors/common.py

```diff
@@ -262,64 +272,12 @@ def _onedal_supported(self, device, method_name, *data):
             return patching_status
         raise RuntimeError(f"Unknown method {method_name} in {class_name}")
 
-    def _kneighbors_dispatch(
-        self, branches, X=None, n_neighbors=None, return_distance=True
-    ):
-        if n_neighbors is None:
-            n_neighbors = self.n_neighbors
-        elif n_neighbors <= 0:
-            raise ValueError("Expected n_neighbors > 0. Got %d" % n_neighbors)
-        else:
-            if not isinstance(n_neighbors, Integral):
-                raise TypeError(
-                    "n_neighbors does not take %s value, "
-                    "enter integer value" % type(n_neighbors)
+    def _fit_queue_check(self, X, queue=None):
+        if X is None and hasattr(self, "_fit_queue"):
+            if queue is None:
+                queue = self._fit_queue
+            elif queue != self._fit_queue:
+                raise RuntimeError(
+                    "Input data shall be located " "on single target device"
```

Comment:
It is weird and unexpected that we have this checker here?
What is KNN case? Why we have this work around and can not to use common approach?

Reply:
This is necessary because I need to preserve the queue used in making any NN fit, the kneighbors method is special as it will reuse the X data if called like 'kneighbors()', so then it is import to know the queue if it is a dpnp or dpctl array/tensor. I was going to preserve the dpnp or dpctl array, but you had said we should use numpy internally, so storing the queue for later use is the only way.

If the following situation were to occur:
```python
q = dpctl.SyclQueue("gpu:0")

X = dpnp.asarray(np.array([[1., 2.], [2., 2.], [2., 3.],
                            [8., 7.], [8., 8.], [25., 80.]], dtype=np.float32),
                            usm_type="device", sycl_queue=q)

est = NearestNeighbors().fit(X)

with config_context(target_offload="gpu:1"):
    est.kneighbors()
```
This should fail. I will need to recheck the queue since I am avoiding checks in _device_offload.dispatch.

I think because of the depth of issues in this I will make a meeting to clarify everything. This is still for testing at this point and will be split into two PRs.